### ISSUE TITLE
cabal-install support window for Cabal library version Custom Setup is built with

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#3861]
Two parts:

1. What should the window be? GHC/Cabal has generally had the property that you need a Setup script built with at least the version of Cabal bundled with GHC to expect things to work. So that would be a reasonable lower bound; Cabal 1.14.0.0 shipped with GHC 7.4.1 so that's probably our window for now.

2. Whatever our support window is, we need to make sure we run cabal-install tests for the sensitive bits on ALL versions of Cabal library. #3790 is a perfect example of a tricky situation that needs tests. So we need tests for 1.14, 1.16, 1.18, 1.20, 1.22 and 1.24.

Seem reasonable?

**Comment by 23Skidoo** [comment]
I think we should decide on a policy for which GHCs we are supporting (for `-w`, not for being built with), and that'll tell us which Cabals we need to support.

**Comment by hvr** [comment]
@23Skidoo it should also be taken into account that the supported range for `-w` will most likely also decide the range of GHCs supported on `matrix.hackage.haskell.org` as I'm not planning to incur the added complexity of supporting/using more than one `cabal-install` version.

**Comment by 23Skidoo** [comment]
@hvr And you want to support GHC versions at least back to 7.0, right?

**Comment by hvr** [comment]
Tbh, I've considered 7.0 to be expendable with `new-build`. However, I'd want to support back till GHC 7.4.2 on `matrix.h.h.o` (which happens to be the first GHC to have proper Generics & SafeHaskell support)

**Comment by ezyang** [comment]
...which suggests that we want 1.14 as our bound for now? ;)

**Comment by ezyang** [comment]
For `new-build`, we have discovered that 1.20 is the first version which supports `--dependency`, which we need in order to put dependencies on the store accurately. So that's the current bound for new-build.

**Comment by hvr** [comment]
@ezyang that effectively means that `new-build` may run into problems with LTS-0/1/2 snapshots?

**Comment by ezyang** [comment]
@hvr Absolutely. Indeed, even with #4058 an early LTS will probably run into trouble because it will try to build the Custom scripts with an old version of Cabal.
